temba/ussd/models.py

```diff
@@ -0,0 +1,98 @@
+from __future__ import absolute_import, unicode_literals
+
+from django.db import models
+from temba.channels.models import ChannelSession
+from temba.contacts.models import Contact, URN
+from temba.triggers.models import Trigger
+
+
+class USSDQuerySet(models.QuerySet):
+    def get(self, *args, **kwargs):
+        kwargs.update(dict(session_type=USSDSession.USSD))
+        return super(USSDQuerySet, self).get(*args, **kwargs)
+
+    def create(self, **kwargs):
+        user = kwargs.get('channel').created_by
+        kwargs.update(dict(session_type=USSDSession.USSD, created_by=user, modified_by=user))
+        return super(USSDQuerySet, self).create(**kwargs)
+
+    def get_initiated_push_session(self, contact):
+        return self.filter(direction=USSDSession.USSD_PUSH, contact=contact).first()
+
+
+class USSDSession(ChannelSession):
+    USSD_PULL = INCOMING = 'I'
+    USSD_PUSH = OUTGOING = 'O'
+
+    objects = USSDQuerySet.as_manager()
+
+    class Meta:
+        proxy = True
+
+    def start_session_async(self):
+        self.flow.start([], [self.contact], start_msg=None, restart_participants=True, session=self)
+
+    def handle_session_async(self, urn, content, date, message_id):
+        from temba.msgs.models import Msg
+
+        message = Msg.create_incoming(channel=self.channel, urn=urn, text=content or '', date=date)
+        message.external_id = message_id
+        message.save()
+
+    def handle_ussd_session_sync(self):
+        # TODO: implement for InfoBip and other sync APIs
+        pass
+
+    @classmethod
+    def handle_incoming(cls, channel, urn, date, external_id, message_id=None, status=None,
+                        flow=None, content=None, starcode=None, org=None, async=True):
+
+        if not external_id or not channel:
+            return False
+
+        trigger = None
+
+        # handle contact with channel
+        urn = URN.from_tel(urn)
+        contact = Contact.get_or_create(channel.org, channel.created_by, urns=[urn], channel=channel)
+        contact_urn = contact.urn_objects[urn]
+
+        contact.set_preferred_channel(channel)
+        contact_urn.update_affinity(channel)
```

Comment:
Curious what we should be doing in the affinity case for USSD since this will also abandon any SMS-based affinity. Is this what we want? Are there likely to ever be two USSD channels in an account? Curious on your thoughts here @nicpottier. I guess it's not unlike voice which does the same thing.